config/custom_dns.go

```diff
@@ -14,14 +14,56 @@ type CustomDNS struct {
 	RewriterConfig      `yaml:",inline"`
 	CustomTTL           Duration         `yaml:"customTTL" default:"1h"`
 	Mapping             CustomDNSMapping `yaml:"mapping"`
+	ZoneFileMapping     ZoneFileDNS      `yaml:"zoneFileMapping" default:""`
 	FilterUnmappedTypes bool             `yaml:"filterUnmappedTypes" default:"true"`
 }
 
 type (
 	CustomDNSMapping map[string]CustomDNSEntries
 	CustomDNSEntries []dns.RR
+
+	ZoneFileDNS CustomDNSMapping
 )
 
+func (z *ZoneFileDNS) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var input string
+	if err := unmarshal(&input); err != nil {
+		return err
+	}
+
+	result := make(ZoneFileDNS, len(input))
+
+	if input != "" {
+		zoneParser := dns.NewZoneParser(strings.NewReader(input), "", "")
+		zoneParser.SetIncludeAllowed(true)
+
+		for {
+			zoneRR, ok := zoneParser.Next()
+
+			if !ok {
+				if zoneParser.Err() != nil {
+					return zoneParser.Err()
+				}
+
+				// Done
+				break
+			}
+
+			domain := zoneRR.Header().Name
+
+			if _, ok := result[domain]; !ok {
+				result[domain] = make(CustomDNSEntries, 0)
```

Comment:
We know there's at least one entry:
```suggestion
				result[domain] = make(CustomDNSEntries, 1)
```